pyscf/pbc/df/incore.py

```diff
@@ -213,3 +135,533 @@ def fill_2c2e(cell, auxcell_or_auxbasis, intor='int2c2e', hermi=0, kpt=numpy.zer
+        Gblksize = min(Gblksize, ngrids, 200000)
+        vGR = vG.real
+        vGI = vG.imag
+
+        for p0, p1 in lib.prange(0, ngrids, Gblksize):
+            # shape of Gpq (nkpts, nGv, nao_pair)
+            Gpq = ft_kern(Gv[p0:p1], gxyz[p0:p1], Gvbase, kpt_allow, kpts)
+            for k, (GpqR, GpqI) in enumerate(zip(*Gpq)):
+                # rho_ij(G) nuc(-G) / G^2
+                # = [Re(rho_ij(G)) + Im(rho_ij(G))*1j] [Re(nuc(G)) - Im(nuc(G))*1j] / G^2
+                vR = np.einsum('k,kx->x', vGR[p0:p1], GpqR)
+                vR+= np.einsum('k,kx->x', vGI[p0:p1], GpqI)
+                if is_zero(kpts[k]):
+                    vj[k] += vR
+                else:
+                    vI = np.einsum('k,kx->x', vGR[p0:p1], GpqI)
+                    vI-= np.einsum('k,kx->x', vGI[p0:p1], GpqR)
+                    vj[k] += vR + vI * 1j
+            Gpq = None
+            t1 = log.timer_debug1('contracting Vnuc [%s:%s]'%(p0, p1), *t1)
+        log.timer_debug1('contracting Vnuc', *t0)
+
+        vj_kpts = []
+        for k, kpt in enumerate(kpts):
+            if is_zero(kpt):
+                vj_kpts.append(lib.unpack_tril(vj[k].real))
+            else:
+                vj_kpts.append(lib.unpack_tril(vj[k]))
+        return np.asarray(vj_kpts)
+
+    def get_nuc(self, mesh=None):
+        t0 = (logger.process_clock(), logger.perf_counter())
+        nuc = self._get_nuc(mesh, with_pseudo=False)
+        logger.timer(self, 'get_nuc', *t0)
+        return nuc
+
+    def get_pp_loc_part1(self, mesh=None):
+        return self._get_nuc(mesh, with_pseudo=True)
+
+    def get_pp_loc_part2(self):
+        cell = self.cell
+        kpts = self.kpts
+        nkpts = len(kpts)
+        intors = ('int3c2e', 'int3c1e', 'int3c1e_r2_origk',
+                  'int3c1e_r4_origk', 'int3c1e_r6_origk')
+        bufR = 0
+        bufI = 0
+        for cn in range(1, 5):
+            fake_cell = pp_int.fake_cell_vloc(cell, cn)
+            if fake_cell.nbas > 0:
+                vR, vI = self.kernel(fake_cell, intors[cn], 's2', comp=1)
+                bufR += np.einsum('...i->...', vR)
+                if vI is not None:
+                    bufI += np.einsum('...i->...', vI)
+
+        if isinstance(bufR, int):
+            if any(cell.atom_symbol(ia) in cell._pseudo for ia in range(cell.natm)):
+                pass
+            else:
+                lib.logger.warn(cell, 'cell.pseudo was specified but its elements %s '
+                                 'were not found in the system.', cell._pseudo.keys())
+            vpploc = [0] * nkpts
+        else:
+            buf = (bufR + bufI * 1j).reshape(nkpts,-1)
+            vpploc = []
+            for k, kpt in enumerate(kpts):
+                v = lib.unpack_tril(buf[k])
+                if abs(kpt).sum() < 1e-9:  # gamma_point:
+                    v = v.real
+                vpploc.append(v)
+        return vpploc
+
+    def get_pp(self):
```

Comment:
Can this function has the same optional argument `mesh=None` for user to overwrite the auto-determined mesh by `_guess_eta`?